File: python/src/ttio/protocols/indexable.py
"""Indexable — O(1) random access by index, key, or range."""
from __future__ import annotations

from typing import Any, Protocol, runtime_checkable


@runtime_checkable
class Indexable(Protocol):
    """Capability for collections that support O(1) random access.

    Every conformer provides index-based access. Key-based and
    range-based access are optional; conformers that do not support
    them raise NotImplementedError.

    Methods
    -------
    object_at_index(index)
        Return the element at index (0-based).
    count()
        Return the total number of elements.
    object_for_key(key)
        Optional. Return the element associated with key or raise
        KeyError.
    objects_in_range(start, stop)
        Optional. Return the half-open slice [start, stop).

    Notes
    -----
    API status: Stable.

    Cross-language equivalents
    --------------------------
    Objective-C: TTIOIndexable ·
    Java: com.dtwthalion.ttio.protocols.Indexable
    """

    def object_at_index(self, index: int) -> Any: ...
    def count(self) -> int: ...

    # Optional members
    def object_for_key(self, key: Any) -> Any: ...
    def objects_in_range(self, start: int, stop: int) -> list[Any]: ...

File: java/src/main/java/com/dtwthalion/ttio/protocols/Indexable.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package com.dtwthalion.ttio.protocols;

import java.util.List;

/**
 * Objects implementing {@code Indexable} support O(1) random access
 * by integer index and, optionally, by key or range. This is the
 * primary access protocol for collections of spectra, runs, and
 * access units.
 *
 * <p>Key-based and range-based access are declared as default
 * methods that throw {@link UnsupportedOperationException};
 * conformers override them if supported.</p>
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOIndexable}, Python
 * {@code ttio.protocols.Indexable}.</p>
 *
 * @param <T> element type
 * @since 0.6
 */
public interface Indexable<T> {

    /** @return the element at {@code index} (0-based). */
    T objectAtIndex(int index);

    /** @return the total number of elements. */
    int count();

    /** Optional. Override to support key-based access. */
    default T objectForKey(Object key) {
        throw new UnsupportedOperationException("objectForKey not supported");
    }

    /** Optional. Override to support range-based access. */
    default List<T> objectsInRange(int start, int stop) {
        throw new UnsupportedOperationException("objectsInRange not supported");
    }
}
